### PR TITLE
Port SAM3 from inference/models to inference-models

### PR DESCRIPTION
**Pull request by grzegorz-roboflow**
## What does this PR do?

Add SAM3 to inference-models

## Type of Change

- New feature (non-breaking change that adds functionality)

## Testing

Added integration tests to cover SAM3

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [x] I have updated the documentation accordingly (if applicable)

## Additional Context

N/A

---

> [!NOTE]
> Introduces SAM3 (Segment Anything 3) to `inference-models` with a PyTorch wrapper and caching utilities.
> 
> - Adds `SAM3Torch` with image embedding, point/box prompting, mask-input refinement, and text/visual prompting APIs; supports batching, logits return, and version validation
> - New in-memory caches for image embeddings and low-res masks with LRU-like eviction; prior-prompt subset lookup optimizes interactive flows
> - Adds entities (`SAM3ImageEmbeddings`, `SAM3Prediction`, `SAM3MaskCacheEntry`) and helpers (prompt serialization, mask selection, hashing)
> - Documentation (`README.md`) with usage examples and a `LICENSE.txt` for SAM3
> - Updates `pyproject.toml` to include `sam3` and platform-conditional `triton` deps
> - Extensive integration tests covering embeddings, segment APIs (points/boxes/masks), text prompting, batching, caching, and error cases
> - Minor fix in legacy `visual_segmentation.py` to short-circuit cache search when prompt has ≤1 point
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bb1a8a2f7d4db5e78fda4b4d84d1ffb5bce3dc8c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: inference_models/inference_models/models/sam3/sam3_torch.py [hunk 1]
```diff
@@ -0,0 +1,925 @@
...
+                np_image = image.cpu().numpy()
+                if np_image.shape[0] == 3:
+                    np_image = np_image.transpose(1, 2, 0)
+                np_image = (
+                    (np_image * 255).astype(np.uint8)
+                    if np_image.max() <= 1
+                    else np_image
+                )
+            else:
+                np_image = image
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference_models/inference_models/models/sam3/sam3_torch.py]: 👍

File: inference_models/inference_models/models/sam3/sam3_torch.py [hunk 2]
```diff
@@ -0,0 +1,925 @@
...
+                SAM3ImageEmbeddings(
+                    image_hash=image_hash,
+                    image_size_hw=size,
+                    embeddings=state,
+                )
+            )
+
+        return result_embeddings
+
+    def segment_images(
```
Comment (PawelPeczek-Roboflow) [inline inference_models/inference_models/models/sam3/sam3_torch.py]: I believe names of methods should follow official nomenclature from sam3 repo & paper - https://arxiv.org/pdf/2511.16719

They use terms: 
* Promptable Visual Segmentation (PVS) for boxes, points masks
* Promptable Concept Segmentation (PCS) for short phrases describing what you are looking for
both are not so easily to translate into method names

maybe then `segment_with_text_prompt(...)` and `segment_with_visual_elements(...)`? not fully committed though

**Review by PawelPeczek-Roboflow — COMMENTED** [review]
generally looks good.

File: inference_models/inference_models/models/sam3/sam3_torch.py (hunk 2 above)
Comment (grzegorz-roboflow) [reply inference_models/inference_models/models/sam3/sam3_torch.py]: Done

File: inference_models/inference_models/models/sam3/sam3_torch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (grzegorz-roboflow) [reply inference_models/inference_models/models/sam3/sam3_torch.py]: Done

File: inference_models/inference_models/models/sam3/cache.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        if self._size_limit is None:
+            return
+        if self._size_limit < 1:
+            raise EnvironmentConfigurationError(
+                message=f"In memory cache size for SAM3 embeddings was set to invalid value. "
+                f"If you are running inference locally - adjust settings of your deployment. If you see this "
+                f"error running on Roboflow platform - contact us to get help.",
+                help_url="https://todo",
+            )
+        while len(self._state) > self._size_limit:
```
[the remark this answers is not in the document]
Comment (grzegorz-roboflow) [reply inference_models/inference_models/models/sam3/cache.py]: Done

**Review by PawelPeczek-Roboflow — DISMISSED** [review]

**Review by PawelPeczek-Roboflow — APPROVED** [review]
